hugr-core/src/export.rs

```diff
@@ -392,6 +406,58 @@ impl<'a> Context<'a> {
         node_id
     }
 
+    /// Export an `OpDef` as an operation declaration.
+    ///
+    /// Operations that allow a declarative form are exported as a reference to
```

Comment:
and those that don't allow this?

Reply:
For now we can fall back to just referencing the operation by its name as before. For each of the `SignatureFunc` variants:

- `PolyFuncType` - This is the only case where we export a declarative op specification at the moment.
- `CustomValidator` - This has a `PolyFuncType` and a custom validator. We can't serialize a custom validator, and so this currently can not be expressed as an op declaration. The idea of the constraint system is to give the custom validator a name so it can be referred to. Once we use constraints for validation, we can handle this case. While we could export by dropping the custom validator, that would not be an accurate description of the operation's type.
- `MissingValidateFunc` - This is like `CustomValidator` except that we do not have the validator code. This case would also be covered by the constraint system, when the operation's type is constrained but we do not have the code that solves the constraint.
- `CustomFunc` - Here we have even less information than for `CustomValidator`; the function that computes the `PolyFuncTypeRV` again can not be serialized. It appears as if the static parameter list must still be known, in which case constraints would allow us to merge this with the `CustomValidator` case.
- `MissingComputeFunc` - Here we know literally nothing about the type of the operation, including what its static parameters are. When we add the `import` mode in addition to `declare` and `define` (i.e. like a declaration except that we only know the name and nothing about its type), we could handle this.

Reply:
have you looked at how these are handled in the current serialization? We use a boolean flag to indicate that some code accompanies this time (either to compute it or to validate it), which allows us to transmit the critical information if we have it while still tracking the original intent. I understand you want to capture all of this in constraints, but it is worth thinking about whether something can be added now that captures this. This would allow this system to replace the existing extension serialization much quicker.

Reply:
I don't see an obvious good solution to doing this with `SignatureFunc` in the current form without implementing some features that we're then likely to remove again (and there's nothing more permanent than a temporary solution). So I propose to postpone the other cases to a future PR.